test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -51,258 +54,518 @@ var (
-	if err != nil {
-		b.Fatal(err)
-	}
-	if err := nodePreparer.PrepareNodes(); err != nil {
-		b.Fatal(err)
-	}
-	defer nodePreparer.CleanupNodes()
+	numPodsScheduledPerNamespace := make(map[string]int)
+	numNodes := 0
 
-	total := 0
-	for _, p := range test.InitPods {
-		if err := createPods(setupNamespace, p, clientset); err != nil {
-			b.Fatal(err)
-		}
-		total += p.Num
-	}
-	if !test.SkipWaitUntilInitPodsScheduled {
-		if err := waitNumPodsScheduled(b, total, podInformer, setupNamespace); err != nil {
+	b.ResetTimer()
+	for opIndex, op := range tc.WorkloadTemplate {
+		realOp, err := op.realOp.patchParams(w)
+		if err != nil {
 			b.Fatal(err)
 		}
-	}
-
-	// start benchmark
-	b.ResetTimer()
-
-	// Start test data collectors.
-	stopCh := make(chan struct{})
-	collectors := getTestDataCollectors(test, podInformer, b)
-	for _, collector := range collectors {
-		go collector.run(stopCh)
-	}
+		switch concreteOp := realOp.(type) {
+		case *createNodesOp:
+			nodePreparer, err := getNodePreparer(fmt.Sprintf("node-%d-", opIndex), concreteOp, clientset)
+			if err != nil {
+				b.Fatal(err)
+			}
+			if err := nodePreparer.PrepareNodes(numNodes); err != nil {
+				b.Fatal(err)
+			}
+			if numNodes == 0 {
+				// Schedule a cleanup at most once. The CleanupNodes function will list
+				// and delete *all* nodes.
+				// TODO(#93794): make CleanupNodes only clean up its own nodes to make
+				// this more intuitive?
+				defer nodePreparer.CleanupNodes()
+			}
+			numNodes += concreteOp.CreateNodes.(int)
+
+		case *createPodsOp:
+			var namespace string
+			if concreteOp.Namespace != nil {
+				namespace = *concreteOp.Namespace
+			} else {
+				namespace = fmt.Sprintf("namespace-%d", opIndex)
+			}
+			var stopCh chan struct{}
+			var collectors []testDataCollector
+			if concreteOp.CollectMetrics {
+				stopCh = make(chan struct{})
+				collectors = getTestDataCollectors(podInformer, fmt.Sprintf("%s/%s", b.Name(), namespace), namespace, tc.MetricsCollectorConfig)
+				for _, collector := range collectors {
+					go collector.run(stopCh)
+				}
+			}
+			if err := createPods(namespace, concreteOp, clientset); err != nil {
+				b.Fatal(err)
+			}
+			barrierAndCollect := func() {
+				if err := barrierOne(podInformer, b.Name(), namespace, concreteOp.CreatePods.(int)); err != nil {
+					b.Fatal(err)
+				}
+				if concreteOp.CollectMetrics {
+					close(stopCh)
+					mu.Lock()
+					for _, collector := range collectors {
+						dataItems = append(dataItems, collector.collect()...)
+					}
+					mu.Unlock()
+				}
+			}
+			if concreteOp.SkipWaitToCompletion {
+				// Only record those namespaces that may potentially require barriers
+				// in the future.
+				if _, ok := numPodsScheduledPerNamespace[namespace]; ok {
+					numPodsScheduledPerNamespace[namespace] += concreteOp.CreatePods.(int)
+				} else {
+					numPodsScheduledPerNamespace[namespace] = concreteOp.CreatePods.(int)
+				}
+				go barrierAndCollect()
```

Comment:
why do we collect in this case? wouldn't metrics get mixed with metrics from future steps?